compiler/util/exprAnalysis.cpp

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright 2004-2016 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "astutil.h"
+#include "baseAST.h"
+#include "stlUtil.h"
+#include "stmt.h"
+#include "exprAnalysis.h"
+
+std::map<Expr*,bool> safeExprCache;
+std::map<FnSymbol*,bool> globalManipFuncCache;
+
+/*
+ * Returns true if `e` has no side effects. Checked side effects are:
+ *  - Read/write to a global
+ *  - Is/containsessential primitive
+ *  - If it's a function call has ref arguments
+ */
+bool exprHasNoSideEffects(Expr* e) {
+  if(safeExprCache.count(e) > 0) {
+    return safeExprCache[e];
+  }
+  if(CallExpr* ce = toCallExpr(e)) {
+    if(!ce->isPrimitive()) {
+      FnSymbol* fnSym = ce->theFnSymbol();
+
+      //const bool cachedGlobalManip = globalManipFuncCache.count(fnSym);
+      const bool cachedGlobalManip = isRegisteredGlobalManip(fnSym);
+
+      if(cachedGlobalManip) {
+        if(globalManipFuncCache[fnSym]) {
+          return false;
+        }
+      }
+      for_formals(formal, fnSym) {
+        if(isReferenceType(formal->typeInfo())) {
+          safeExprCache[e] = false;
+          return false;
+        }
+      }
+
+      std::vector<BaseAST*> asts;
+      collect_asts(fnSym->body, asts);
+      if(!cachedGlobalManip){
+        for_vector(BaseAST, ast, asts) {
+          if (SymExpr* se = toSymExpr(ast)) {
+            Symbol* var = se->var;
+
+            if(!var->isImmediate() &&  isGlobal(var)){
+              safeExprCache[e] = false;
+              globalManipFuncCache[fnSym] = true;
+              return false;
+            }
+          }
+        }
+      }
+
+      //this else if lazily marks a def to be immovable if the
+      //function body has a call to another user function
+      //this can be enhanced by going recursive, but it might be a bit overkill
+      for_vector(BaseAST, ast, asts) {
+        if(CallExpr* ce = toCallExpr(ast)) {
+          if(!isNonEssentialPrimitive(ce)) {
+            safeExprCache[e] = false;
+            return false;
+          }
+        }
+      }
+      //it shouldn't be touching any globals at this point
+      globalManipFuncCache[fnSym] = false;
```

Comment:
Generally speaking, I think it'd be nice to put all of this analysis for whether or not a FnSymbol is safe into a separate function. So the above block would become:

      if(cachedGlobalManip) {
        bool fnHasSideEffects = globalManipFuncCache[fnSym];
        return !fnHasSideEffects;
      }

      bool fnHasSideEffects = determineIfFnHasSideEffects(fnSym);
      globalManipFuncCache[fnSym] = fnHasSideEffects;
      return fnHasSideEffects;